src/App.js

```diff
@@ -29,37 +30,103 @@ const App = () => {
+  // const [gameEnd, setGameEnd] = useState(false);
 
-  // Wave 2
+  const squareClicked =(id)=>{
+    let didPlay = false 
+    if (winner !== null) {return}
+    const newState = squares.map(row =>row.map(pos =>{
+      if (pos.id !== id) {return pos;}
+      if (pos.value !== '') {return pos;} 
+  //we know the id matches
+      didPlay = true
+      const newPos = {...pos}
+      newPos.value = currentPlayer
+      return newPos 
+    }))
+    if (didPlay) {
+      setSquares(newState) 
+      if (currentPlayer === PLAYER_1) {
+        setCurrentPlayer(PLAYER_2)
+      } else {setCurrentPlayer(PLAYER_1)}
+    }
+  };
+  
+  // let winner = checkForWinner();
+  // console.log(winner);
+  // if (winner !== null){
+  //   setWinner(winner);
+  //   setGameEnd(true);
+    
+    // break;
+  // }
+  // // Wave 2
   // You will need to create a method to change the square
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
-
+ 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    let i = 0;
+
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
   };
 
   const resetGame = () => {
-    // Complete in Wave 4
+    setSquares(generateSquares());
+    setCurrentPlayer(PLAYER_1);
+    // setWinner(null);
+    // setGameEnd(false);
   };
 
+  const winner = checkForWinner()
+  // console.log(winner)
+
   return (
-    <div className="App">
-      <header className="App-header">
+    <div className='App'>
+      <header className='App-header'>
         <h1>React Tic Tac Toe</h1>
-        <h2>The winner is ... -- Fill in for wave 3 </h2>
-        <button>Reset Game</button>
+        <h2>Winner is {winner} </h2>
```

Comment:
It would be nice to use conditional rendering here to avoid displaying the winner message until after the game is over. Alternatively, we could display the current player's turn here until the game has ended. We could also think about how to detect whether the game ended in a tie, and display a message to that effect.